### PR TITLE
Persistent index trigger compaction

### LINKED ISSUE #3929: Triger PersistetIndex flush or merge compaction when conditions met
Currently, PersistetIndex will only do snapshotting and appending WAL, which will let memory increase unbounded, to limit memory usage, flush and merge compaction should be triggered when certain conditions met.

### PR DESCRIPTION
**Pull request by sevev**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [x] others

## Which issues of this PR fixes ：

Fixes #3929

## Problem Summary(Required) ：

Currently, PersistetIndex will only do snapshotting and appending WAL, which will let memory increase unbounded, to limit memory usage, flush and merge compaction should be triggered when certain conditions met.

The trigger conditions are below:
1. L0_mem_size > L0_flush_size_min
2. L0_mem_size > L0_snapshot_size_max &&  L1_file_size / L0_mem_size < l0_l1_merge_ratio

If L1 is nullptr, do `flush_l0`, otherwise do `merge_compaction`.

### CODE REVIEW

**Comment by sevev** [comment]
run starrocks_be_unittest

File: be/src/storage/persistent_index.cpp [hunk 1]
```diff
@@ -1375,21 +1466,27 @@ Status PersistentIndex::_merge_compaction() {
...
-                                            estimated_size_per_shard, kvs));
-            RETURN_IF_ERROR(writer.write_shard(_key_size, value_size, npage_hint, kvs));
-            // clear to optimize memory usage
-            l0_kvs_by_shard[cur_shard_idx].clear();
-            l0_kvs_by_shard[cur_shard_idx].shrink_to_fit();
-            l1_kvs_by_shard[cur_shard_idx].clear();
-            l1_kvs_by_shard[cur_shard_idx].shrink_to_fit();
-            cur_shard_idx++;
+        if (cur_shard_idx < num_shard_finished) {
+            while (cur_shard_idx < num_shard_finished) {
```
Comment (decster) [inline be/src/storage/persistent_index.cpp:1477]: those conditions are identical, maybe it's redundant

File: be/src/storage/persistent_index.cpp [hunk 2]
```diff
@@ -1216,6 +1254,56 @@ Status PersistentIndex::_flush_l0() {
...
+// do _flush_l0 or merge compaction.
+Status PersistentIndex::_check_and_flush_l0() {
+    size_t kv_size = _key_size + sizeof(IndexValue);
+    size_t l0_mem_size = kv_size * _l0->size();
+    uint64_t l1_file_size = 0;
+    if (_l1 != nullptr) {
+        _l1->file_size(&l1_file_size);
+    }
+    if (l0_mem_size <= l0_flush_size_min &&
+        ((l0_mem_size <= l0_snapshot_size_max) || (l1_file_size / l0_mem_size > 10))) {
```
Comment (decster) [inline be/src/storage/persistent_index.cpp:1293]: 10 -> const var defined earlier

File: be/src/storage/persistent_index.cpp [hunk 3]
```diff
@@ -1216,6 +1254,56 @@ Status PersistentIndex::_flush_l0() {
...
+        return Status::OK();
+    }
+    // flush _l0
+    if (_l1 == nullptr) {
+        RETURN_IF_ERROR(_flush_l0());
+    } else {
+        RETURN_IF_ERROR(_merge_compaction());
+    }
+    // rebuild _l0 and _l1
+    RETURN_IF_ERROR(_rebuild());
```
Comment (decster) [inline be/src/storage/persistent_index.cpp:1303]: I think the more clear order is:
commit:

1. check condition, chose which action to do: snaphot/wal/flush/merge_compaction
2. update meta according to different action

on_commit:

if necessary, e.g. flush/merge_compaction occurred
    * reload(_rebuild) from new versions meta&file
    * clear old files

File: be/src/storage/persistent_index.cpp [hunk 4]
```diff
@@ -49,10 +49,19 @@ T pad(T v, P p) {
     return npad(v, p) * p;
 }
 
+uint32_t log2(size_t n) {
```
Comment (chaoyli) [inline be/src/storage/persistent_index.cpp:52]: util/bit_util.h have this implementation. You can use the util instead.

File: be/src/storage/persistent_index.cpp [hunk 5]
```diff
@@ -49,10 +49,19 @@ T pad(T v, P p) {
...
+        x++;
+    }
+    return x;
+}
+
 struct IndexHash {
     IndexHash() {}
     IndexHash(uint64_t hash) : hash(hash) {}
-    uint64_t shard() const { return hash >> 48; }
+    uint64_t shard(uint32_t n) const { return (n == 0) ? 0 : hash >> (64 - n); }
```
[the remark this answers is not in the document]
Comment (sevev) [reply be/src/storage/persistent_index.cpp]: ISO 9899:2011 6.5.7 Bitwise shift operators

`If the value of the right operand is negative or is greater than or equal to the width of the promoted left operand, the behavior is undefined.`

when `n` is 0,  the result of `hash >> (64 -n)` is equal to hash but not 0, so use `(hash >> (63 - n)) >> 1` to handle it

File: be/src/storage/persistent_index.cpp (hunk 3 above)
Comment (sevev) [reply be/src/storage/persistent_index.cpp]: > I think the more clear order is: commit:
> 
> 1. check condition, chose which action to do: snaphot/wal/flush/merge_compaction
> 2. update meta according to different action
> 
> on_commit:
> 
> if necessary, e.g. flush/merge_compaction occurred * reload(_rebuild) from new versions meta&file * clear old files

done, change in https://github.com/StarRocks/starrocks/pull/4172/commits/19bb730532d645dd8192ac70522ad282327f364c

File: be/src/storage/persistent_index.cpp (hunk 4 above)
Comment (sevev) [reply be/src/storage/persistent_index.cpp]: > util/bit_util.h have this implementation. You can use the util instead.

done, change in https://github.com/StarRocks/starrocks/pull/4172/commits/19bb730532d645dd8192ac70522ad282327f364c

**Comment by sevev** [comment]
run all

**Review by decster — DISMISSED** [review]

**Comment by sevev** [comment]
run all

**Comment by sevev** [comment]
run starrocks_fe_unittest

**Comment by sevev** [comment]
run starrocks_fe_unittest

**Comment by sevev** [comment]
run starrocks_fe_unittest

**Comment by sevev** [comment]
run starrocks_fe_unittest

**Review by decster — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**